# Data Analysis with Pandas

In this notebook, I will be using Pandas to read Adult dataset and to perform some basic analysis to improve the understanding of the dataset.This Notebook demonstrates various processes of data analysis step-by-step.

### Goals:  
- Program in Python using Jupyter notebook
- Perform Data Analysis using Pandas
- Practice data pre-processing methods
- Analyze and summarize the dataset by finding facts from the data

#### Dataset:  
Adult - https://archive.ics.uci.edu/ml/datasets/Adult (https://archive.ics.uci.edu/ml/datasets/Adult)
- Please read Adult webpage carefully including Attribute Information section to familiarise yourself with the data and the data structure.
- To download data, click 'Data Folder' and select 'adult.data'. Save the data file as .csv file.
- Attributes: You will also need to see 'adult.names' for the attribute names. Insert a row at the top of the dataset and add attribute names to the respective columns.
- You will notice that the last column has no name. Name the last column as 'class-label'.


## Exploratory analysis: Loading and exploring the dataset

In this section we are going to gain more useful insights of the dataset. The first step is to read the dataset into a dataframe and perform basic analysis to gain more understanding of the dataset we are working on such as what are the maximum and minimum values in the dataset, whether there are NULL values that exist in the dataset,summarizing data using 'groupby' etc

In [29]:
#importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [30]:
#loading the dataset to a dataframe
data = pd.read_csv('C:/Users/25471/Desktop/DataMining_projects/datamining/adult.csv')

In [31]:
data.head()

age          workclass  fblwgt   education  education-num   
0   39          State-gov   77516   Bachelors             13  \
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital status          occupation    relationship    race      sex   
0        Never-married        Adm-clerical   Not-in-family   White     Male  \
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital gain  capital loss  hours per week  native country class label  
0          2174             0              40   United-States       <=50K  
1             0             0              13   United-States       <=50K  
2             0             0              40   United-States       <=50K  
3             0             0              40   United-States       <=50K  
4             0             0              40            Cuba       <=50K

Here data.head() is used to display and examine the first five rows of the dataframe data.
The output shows a demographic respresentation of individuals.


- age: The age of each individual. It ranges from 28 to 53 in the given sample.
- workclass: The type of work class or employment status, such as "State-gov," "Self-emp-not-inc," and "Private."
- fblwgt: This column's meaning is unclear.
- education and education-num: These columns provide information about the individual's educational background. It includes values like "Bachelors," "HS-grad," and "11th." The education-num column represents the numerical representation of education levels.
- marital status: Describes the marital status of individuals, including categories like "Never-married," "Married-civ-spouse," and "Divorced."
- occupation: Indicates the occupation of each individual, such as "Adm-clerical," "Exec-managerial," and "Handlers-cleaners."
- relationship: Defines the relationship status of the individuals, such as "Not-in-family," "Husband," and "Wife."
- race: Specifies the racial background of the individuals, including categories like "White," "Black," and potentially others.
- sex: Indicates the gender of each individual, either "Male" or "Female."
- capital gain and capital loss: These columns provide information about capital gains and losses, although all the given samples have zero values in both columns.
- hours per week: Represents the number of hours worked per week, which is 40 for all the given samples.
- native country: Specifies the native country of the individuals, such as "United-States" and "Cuba."
- class label: Indicates the class label or income level, with the values "<=50K" suggesting an income less than or equal to 50,000 units.

#### Q1. Use head(2), head(10), tail(2). Explain your observations

In [32]:
data.head(2)

age          workclass  fblwgt   education  education-num   
0   39          State-gov   77516   Bachelors             13  \
1   50   Self-emp-not-inc   83311   Bachelors             13   

        marital status        occupation    relationship    race    sex   
0        Never-married      Adm-clerical   Not-in-family   White   Male  \
1   Married-civ-spouse   Exec-managerial         Husband   White   Male   

   capital gain  capital loss  hours per week  native country class label  
0          2174             0              40   United-States       <=50K  
1             0             0              13   United-States       <=50K

In [33]:
data.head(10)

age          workclass  fblwgt   education  education-num   
0   39          State-gov   77516   Bachelors             13  \
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   
5   37            Private  284582     Masters             14   
6   49            Private  160187         9th              5   
7   52   Self-emp-not-inc  209642     HS-grad              9   
8   31            Private   45781     Masters             14   
9   42            Private  159449   Bachelors             13   

           marital status          occupation    relationship    race   
0           Never-married        Adm-clerical   Not-in-family   White  \
1      Married-civ-spouse     Exec-managerial         Husband   White   
2                Divorced   Handlers-cleaners   Not-in-family   White   
3      Married-civ-spouse   Handlers-cleaners         Husband   Black   
4      Married-civ-spouse      Prof-specialty            Wife   Black   
5      Married-civ-spouse     Exec-managerial            Wife   White   
6   Married-spouse-absent       Other-service   Not-in-family   Black   
7      Married-civ-spouse     Exec-managerial         Husband   White   
8           Never-married      Prof-specialty   Not-in-family   White   
9      Married-civ-spouse     Exec-managerial         Husband   White   

       sex  capital gain  capital loss  hours per week  native country   
0     Male          2174             0              40   United-States  \
1     Male             0             0              13   United-States   
2     Male             0             0              40   United-States   
3     Male             0             0              40   United-States   
4   Female             0             0              40            Cuba   
5   Female             0             0              40   United-States   
6   Female             0             0              16         Jamaica   
7     Male             0             0              45   United-States   
8   Female         14084             0              50   United-States   
9     Male          5178             0              40   United-States   

  class label  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
5       <=50K  
6       <=50K  
7        >50K  
8        >50K  
9        >50K

In [34]:
data.tail(2)

age      workclass  fblwgt education  education-num   
32559   22        Private  201490   HS-grad              9  \
32560   52   Self-emp-inc  287927   HS-grad              9   

            marital status        occupation relationship    race      sex   
32559        Never-married      Adm-clerical    Own-child   White     Male  \
32560   Married-civ-spouse   Exec-managerial         Wife   White   Female   

       capital gain  capital loss  hours per week  native country class label  
32559             0             0              20   United-States       <=50K  
32560         15024             0              40   United-States        >50K

- Here head(2) displays the first two rows of the dataframe data.
- head(10) returns the first 10 rows of the dataframe.
- tail(2) displays the last two rows of the dataframe.

In [35]:
data.shape

(32561, 15)

data.shape is used to get the dimensionality of the dataframe data.This returns the number of rows (32561) and columns (15)

## Generating a unique dataset 
For this task we are going to generate a our version of dataset. To achieve this, we will use 48 in random_state.

In [36]:
data = data.sample(n=30000, random_state = 48)

The sample() function is used to randomly select rows from a DataFrame.It selects a random sample of 30,000 rows from the dataset. 
The random sample will consistently provide the same results when the code is run again with the same seed if a specific value, in this example 48, is specified, hence ensures consistency.

In [37]:
data.shape

(30000, 15)

The dimensionality of the dataset has changed in number of rows from 32561 to 30000 with no changes in number of columns

In [38]:
data.describe()

age        fblwgt  education-num  capital gain  capital loss   
count  30000.000000  3.000000e+04   30000.000000  30000.000000  30000.000000  \
mean      38.601733  1.897456e+05      10.078533   1072.036100     86.703967   
std       13.645910  1.058005e+05       2.571715   7387.730029    401.942288   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.177670e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783090e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.368742e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours per week  
count    30000.000000  
mean        40.409433  
std         12.349540  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

This method displays a summary of several statistical indicators for each numerical column, including count, mean, standard deviation, minimum value, quartiles, and maximum value.

- count: Shows how many non-missing values there are in each column. In this instance, all columns have 30,000 non-missing data.

- mean: Displays each column's average or mean value. For instance, the typical age is about 38.60 and so on.

- std: stands for the standard deviation, which quantifies how much results deviate from the mean. More variety in the data is indicated by a higher standard deviation.

- min: gives the minimum value that was noticed in each column, or min. For instance,the minimum age is 17.

- 25%, 50%, and 75%: These are quartiles,25% of the data falls is known as the 25th percentile (first quartile). Similar to this, the median is represented by the 50th percentile (second quartile), and the number below which 75% of the data falls by the 75th percentile (third quartile).

- max: Displays the highest value ever recorded for each column. For instance, the maximum age is 90.


In [39]:
data['education-num'].value_counts()

education-num
9     9678
10    6724
13    4907
14    1600
11    1271
7     1091
12     985
6      869
4      589
15     526
5      471
8      401
16     381
3      301
2      157
1       49
Name: count, dtype: int64

This returns the occurence of unique values in the 'education-num' column of integer type.Each value in the 'education-num' column represents a different level of education or educational attainment for example 9 occurs 9678 times and so on.

In [40]:
data['education'].value_counts()

education
 HS-grad         9678
 Some-college    6724
 Bachelors       4907
 Masters         1600
 Assoc-voc       1271
 11th            1091
 Assoc-acdm       985
 10th             869
 7th-8th          589
 Prof-school      526
 9th              471
 12th             401
 Doctorate        381
 5th-6th          301
 1st-4th          157
 Preschool         49
Name: count, dtype: int64

This returns the occurence of each value in the level of education in the 'education' column. For example HS-grad has 9678 graduates and so on.

In [41]:
data = data.drop(['fblwgt'], axis=1)

The cell above removes/drops the column with the label "fblwgt" by specifying the axis=1 for columns and axis=0 for rows.
We drop this column because it has no role in our exploration.

In [15]:
data.shape

(30000, 14)

Change in the number of columns from 15 to 14